File: src/main/java/com/thebluehats/server/core/ApplicationBuilder.java
package com.thebluehats.server.core;

import com.google.inject.Guice;
import com.google.inject.Injector;
import com.thebluehats.server.core.modules.*;
import com.thebluehats.server.core.services.Service;

import java.util.ArrayList;

public class ApplicationBuilder {
    private final ArrayList<Class<? extends Service>> services = new ArrayList<>();

    private final TheBlueHatsServerPlugin plugin;

    public ApplicationBuilder(TheBlueHatsServerPlugin plugin) {
        this.plugin = plugin;
    }

    public ApplicationBuilder addService(Class<? extends Service> clazz) {
        services.add(clazz);

        return this;
    }

    public Application build() {
        Injector injector = provisionInjector();

        for (Class<? extends Service> service : services) {
            injector.getInstance(service).provision(injector);
        }

        return injector.getInstance(Application.class);
    }

    private Injector provisionInjector() {
        return Guice.createInjector(new PluginModule(plugin), new RegionManagerModule(),
                new CustomEnchantManagerModule(), new CombatManagerModule(), new EventVerifiersModule(),
                new DamageManagerModule(), new BowManagerModule(), new TimerModule(), new HitCounterModule(),
                new MirrorModule(), new CustomEnchantUtilsModule(), new ServerAPIModule(),
                new PitDataRepositoryModule(), new RomanNumeralConverterModule(), new PantsDataContainerModule(),
                new PerformanceStatsServiceModule(), new GlobalTimerModule(), new DamageEnchantTriggersModule(),
                new RegistererModule(), new PitScoreboardModule());
    }
}

File: src/main/java/com/thebluehats/server/core/ApplicationBuilder.kt
package com.thebluehats.server.core

import com.google.inject.Guice
import com.google.inject.Injector
import com.thebluehats.server.core.modules.*
import com.thebluehats.server.core.services.Service
import java.util.*

class ApplicationBuilder(private val plugin: TheBlueHatsServerPlugin) {
    private val services = ArrayList<Class<out Service?>>()

    fun addService(clazz: Class<out Service?>): ApplicationBuilder {
        services.add(clazz)
        return this
    }

    fun build(): Application {
        val injector = provisionInjector()

        for (service in services) {
            injector.getInstance(service)!!.provision(injector)
        }

        return injector.getInstance(Application::class.java)
    }

    private fun provisionInjector(): Injector {
        return Guice.createInjector(
            PluginModule(plugin), RegionManagerModule(),
            CustomEnchantManagerModule(), CombatManagerModule(), EventVerifiersModule(),
            DamageManagerModule(), BowManagerModule(), TimerModule(), HitCounterModule(),
            MirrorModule(), CustomEnchantUtilsModule(), ServerAPIModule(),
            RomanNumeralConverterModule(), PantsDataContainerModule(),
            PerformanceStatsServiceModule(), GlobalTimerModule(), DamageEnchantTriggersModule(),
            RegistererModule(), PitScoreboardModule()
        )
    }
}